File: src/priority_queue.py
"""Deterministic, explainable ordering for a hospital triage queue."""

from datetime import datetime, timezone


def order_patients(patients, now=None):
    """Return patients ordered by triage tier, priority score, then arrival time.

    Emergency patients always remain ahead of Normal patients. A Normal patient
    waiting at least 60 minutes receives a +10 escalation boost.
    """
    if now is None:
        now = datetime.now(timezone.utc)

    scored_patients = []
    for patient in patients:
        if patient["triage"] not in {"EMERGENCY", "NORMAL"}:
            raise ValueError("triage must be EMERGENCY or NORMAL")
        severity = patient["severity"]
        if not isinstance(severity, int) or not 1 <= severity <= 10:
            raise ValueError("severity must be an integer from 1 to 10")

        waited_minutes = max(0, int((now - patient["arrived_at"]).total_seconds() // 60))
        escalation_boost = 10 if patient["triage"] == "NORMAL" and waited_minutes >= 60 else 0
        scored_patients.append((patient, severity + escalation_boost))

    return [
        patient
        for patient, _score in sorted(
            scored_patients,
            key=lambda item: (
                0 if item[0]["triage"] == "EMERGENCY" else 1,
                -item[1],
                item[0]["arrived_at"],
            ),
        )
    ]


def get_urgent_patient(queue, current_time=None):
    """Return the next patient to treat, or None when the queue is empty."""
    ordered_queue = order_patients(queue, current_time)
    return ordered_queue[0] if ordered_queue else None

File: src/priorityQueue.mjs
/**
 * Order the triage queue without ever allowing a normal case to overtake an
 * emergency case. A normal case waiting at least 60 minutes receives an
 * escalation boost so it is considered ahead of newly-arrived normal cases.
 * Earlier arrival wins an otherwise equal score.
 */
export function orderPatients(patients, now = new Date()) {
  if (!Array.isArray(patients)) throw new TypeError('patients must be an array');
  const nowMs = new Date(now).getTime();
  if (Number.isNaN(nowMs)) throw new TypeError('now must be a valid date');

  return patients
    .map((patient) => {
      if (!['EMERGENCY', 'NORMAL'].includes(patient.triage)) {
        throw new RangeError('triage must be EMERGENCY or NORMAL');
      }
      if (!Number.isInteger(patient.severity) || patient.severity < 1 || patient.severity > 10) {
        throw new RangeError('severity must be an integer from 1 to 10');
      }
      const arrivedMs = new Date(patient.arrivedAt).getTime();
      if (Number.isNaN(arrivedMs)) throw new TypeError('arrivedAt must be a valid date');
      const waitedMinutes = Math.max(0, Math.floor((nowMs - arrivedMs) / 60_000));
      return {
        ...patient,
        priority: patient.severity + (patient.triage === 'NORMAL' && waitedMinutes >= 60 ? 10 : 0),
        arrivedMs,
      };
    })
    .sort((a, b) => {
      const triageDifference = Number(b.triage === 'EMERGENCY') - Number(a.triage === 'EMERGENCY');
      if (triageDifference !== 0) return triageDifference;
      if (a.priority !== b.priority) return b.priority - a.priority;
      return a.arrivedMs - b.arrivedMs;
    })
    .map(({ arrivedMs, ...patient }) => patient);
}
